File: app/src/main/java/com/qwe7002/telegram_rc/notification_listener_service.java
package com.qwe7002.telegram_rc;

import android.app.Notification;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.IBinder;
import android.service.notification.NotificationListenerService;
import android.service.notification.StatusBarNotification;
import android.util.Log;

import androidx.annotation.NonNull;

import com.google.gson.Gson;
import com.qwe7002.telegram_rc.config.proxy;
import com.qwe7002.telegram_rc.data_structure.request_message;
import com.qwe7002.telegram_rc.static_class.const_value;
import com.qwe7002.telegram_rc.static_class.log_func;
import com.qwe7002.telegram_rc.static_class.network_func;
import com.qwe7002.telegram_rc.static_class.other_func;
import com.qwe7002.telegram_rc.static_class.resend_func;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import io.paperdb.Paper;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class notification_listener_service extends NotificationListenerService {
    final String TAG = "notification_receiver";
    static Map<String, String> app_name_list = new HashMap<>();
    Context context;
    SharedPreferences sharedPreferences;
    @Override
    public void onCreate() {
        super.onCreate();
        context = getApplicationContext();
        Paper.init(context);
        sharedPreferences = context.getSharedPreferences("data", Context.MODE_PRIVATE);
        Notification notification = other_func.get_notification_obj(getApplicationContext(), getString(R.string.Notification_Listener_title));
        startForeground(com.qwe7002.telegram_rc.notify_id.NOTIFICATION_LISTENER_SERVICE, notification);

    }

    @Override
    public void onDestroy() {
        stopForeground(true);
        super.onDestroy();
    }

    @Override
    public void onNotificationPosted(@NotNull StatusBarNotification sbn) {
        final String package_name = sbn.getPackageName();
        Log.d(TAG, "onNotificationPosted: " + package_name);

        if (!sharedPreferences.getBoolean("initialized", false)) {
            Log.i(TAG, "Uninitialized, Notification receiver is deactivated.");
            return;
        }

        List<String> listen_list = Paper.book("system_config").read("notify_listen_list", new ArrayList<>());
        if (!listen_list.contains(package_name)) {
            Log.i(TAG, "[" + package_name + "] Not in the list of listening packages.");
            return;
        }
        Bundle extras = sbn.getNotification().extras;
        assert extras != null;
        String app_name = "unknown";
        Log.d(TAG, "onNotificationPosted: " + app_name_list);
        if (app_name_list.containsKey(package_name)) {
            app_name = app_name_list.get(package_name);
        } else {
            final PackageManager pm = getApplicationContext().getPackageManager();
            try {
                ApplicationInfo application_info = pm.getApplicationInfo(sbn.getPackageName(), 0);
                app_name = (String) pm.getApplicationLabel(application_info);
                app_name_list.put(package_name, app_name);
            } catch (PackageManager.NameNotFoundException e) {
                e.printStackTrace();
            }
        }
        String title = extras.getString(Notification.EXTRA_TITLE, "None");
        String content = extras.getString(Notification.EXTRA_TEXT, "None");

        String bot_token = sharedPreferences.getString("bot_token", "");
        String chat_id = sharedPreferences.getString("chat_id", "");
        String request_uri = network_func.get_url(bot_token, "sendMessage");
        request_message request_body = new request_message();
        request_body.chat_id = chat_id;
        request_body.message_thread_id = sharedPreferences.getString("message_thread_id", "");
        request_body.text = getString(R.string.receive_notification_title) + "\n" + getString(R.string.app_name_title) + app_name + "\n" + getString(R.string.title) + title + "\n" + getString(R.string.content) + content;
        RequestBody body = RequestBody.create(new Gson().toJson(request_body), const_value.JSON);
        OkHttpClient okhttp_client = network_func.get_okhttp_obj(sharedPreferences.getBoolean("doh_switch", true), Paper.book("system_config").read("proxy_config", new proxy()));
        Request request = new Request.Builder().url(request_uri).method("POST", body).build();
        Call call = okhttp_client.newCall(request);
        final String error_head = "Send notification failed:";
        call.enqueue(new Callback() {
            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {
                e.printStackTrace();
                log_func.write_log(context, error_head + e.getMessage());
                resend_func.add_resend_loop(context, request_body.text);
            }

            @Override
            public void onResponse(@NonNull Call call, @NonNull Response response) throws IOException {
                assert response.body() != null;
                String result = Objects.requireNonNull(response.body()).string();
                if (response.code() != 200) {
                    log_func.write_log(context, error_head + response.code() + " " + result);
                    resend_func.add_resend_loop(context, request_body.text);
                }
            }
        });
    }

    @Override
    public IBinder onBind(Intent intent) {
        return super.onBind(intent);
    }


}

File: app/src/main/java/com/qwe7002/telegram_rc/NotificationListenerService.kt
package com.qwe7002.telegram_rc

import android.app.Notification
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.content.pm.ServiceInfo
import android.os.IBinder
import android.service.notification.NotificationListenerService
import android.service.notification.StatusBarNotification
import android.util.Log
import com.google.gson.Gson
import com.qwe7002.telegram_rc.data_structure.request_message
import com.qwe7002.telegram_rc.static_class.CONST
import com.qwe7002.telegram_rc.static_class.log
import com.qwe7002.telegram_rc.static_class.network
import com.qwe7002.telegram_rc.static_class.notify
import com.qwe7002.telegram_rc.static_class.other
import com.qwe7002.telegram_rc.static_class.resend
import io.paperdb.Paper
import okhttp3.Call
import okhttp3.Callback
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.io.IOException
import java.util.Objects

class NotificationListenerService : NotificationListenerService() {
    val TAG: String = "notification_receiver"
    lateinit var  sharedPreferences: SharedPreferences

    override fun onCreate() {
        super.onCreate()
        Paper.init(applicationContext)
        sharedPreferences = applicationContext.getSharedPreferences("data", MODE_PRIVATE)
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {

        startForegroundNotification()
        return START_STICKY
    }

    fun startForegroundNotification() {
        val notification: Notification.Builder = other.getNotificationObj(
            applicationContext,
            getString(R.string.Notification_Listener_title)
        )
        // Create a PendingIntent for the broadcast receiver
        val deleteIntent = Intent(applicationContext, DeleteReceiver::class.java)
        val pendingIntent = PendingIntent.getBroadcast(
            applicationContext,
            notify.NOTIFICATION_LISTENER_SERVICE,
            deleteIntent,
            PendingIntent.FLAG_CANCEL_CURRENT
        )
        // Set the deleteIntent on the notification
        notification.setDeleteIntent(pendingIntent)
        startForeground(
            notify.NOTIFICATION_LISTENER_SERVICE, notification.build(),
            ServiceInfo.FOREGROUND_SERVICE_TYPE_LOCATION
        )
    }

    internal inner class DeleteReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            // Get the notification ID from the intent
            Log.d("battery", "onReceive: Received notification that it was removed, try to pull it up again.")
            val notificationId = intent.getIntExtra(Notification.EXTRA_NOTIFICATION_ID, 0)
            if (notificationId == notify.BATTERY) {
                startForegroundNotification()
            }
        }
    }

    override fun onNotificationPosted(sbn: StatusBarNotification) {
        val package_name = sbn.packageName
        Log.d(TAG, "onNotificationPosted: $package_name")

        if (!sharedPreferences.getBoolean("initialized", false)) {
            Log.i(TAG, "Uninitialized, Notification receiver is deactivated.")
            return
        }

        val listenList: List<String> =
            Paper.book("system_config").read("notify_listen_list", ArrayList())!!
        if (!listenList.contains(package_name)) {
            Log.i(TAG, "[$package_name] Not in the list of listening packages.")
            return
        }
        val extras = sbn.notification.extras!!
        var appName: String? = "unknown"
        Log.d(TAG, "onNotificationPosted: $appNameList")
        if (appNameList.containsKey(package_name)) {
            appName = appNameList[package_name]
        } else {
            val pm = applicationContext.packageManager
            try {
                val application_info = pm.getApplicationInfo(sbn.packageName, 0)
                appName = pm.getApplicationLabel(application_info) as String
                appNameList[package_name] = appName
            } catch (e: PackageManager.NameNotFoundException) {
                e.printStackTrace()
            }
        }
        val title = extras.getString(Notification.EXTRA_TITLE, "None")
        val content = extras.getString(Notification.EXTRA_TEXT, "None")
        if (title == "None" && content == "None") {
            return
        }
        val botToken = sharedPreferences.getString("bot_token", "")
        val chatId = sharedPreferences.getString("chat_id", "")
        val requestUri = network.getUrl(botToken, "sendMessage")
        val requestBody = request_message()
        if ((System.currentTimeMillis() - lastSendTime) <= 1000L && (lastPackage == package_name)) {
            if (lastMessage == title + content) {
                return
            }
        }
        requestBody.chat_id = chatId
        requestBody.message_thread_id = sharedPreferences.getString("message_thread_id", "")
        requestBody.text = """
            ${getString(R.string.receive_notification_title)}
            ${getString(R.string.app_name_title)}$appName
            ${getString(R.string.title)}$title
            ${getString(R.string.content)}$content
            """.trimIndent()
        val body: RequestBody = Gson().toJson(requestBody).toRequestBody(CONST.JSON)
        val okhttpClient = network.getOkhttpObj(
            sharedPreferences.getBoolean("doh_switch", true)
        )
        val request: Request = Request.Builder().url(requestUri).method("POST", body).build()
        val call = okhttpClient.newCall(request)
        lastPackage = package_name
        lastMessage = title + content
        lastSendTime = System.currentTimeMillis()
        val error_head = "Send notification failed:"
        call.enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                e.printStackTrace()
                log.writeLog(applicationContext, error_head + e.message)
                resend.addResendLoop(applicationContext, requestBody.text)
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                val result = Objects.requireNonNull(response.body).string()
                if (response.code != 200) {
                    log.writeLog(applicationContext, error_head + response.code + " " + result)
                    resend.addResendLoop(applicationContext, requestBody.text)
                }
            }
        })
    }

    override fun onBind(intent: Intent): IBinder? {
        return super.onBind(intent)
    }


    companion object {
        var appNameList: MutableMap<String, String?> = HashMap()
        var lastPackage: String? = null
        var lastMessage: String? = null
        var lastSendTime: Long = 0
    }
}
